src/iface/ethernet.rs

```diff
@@ -828,11 +833,11 @@ impl<'b, 'c> InterfaceInner<'b, 'c> {
 
     #[cfg(feature = "proto-ipv4")]
     fn process_icmpv4<'frame>(&self, _sockets: &mut SocketSet, ip_repr: IpRepr,
-                              ip_payload: &'frame [u8]) -> Result<Packet<'frame>>
+                            ip_payload: &'frame [u8]) -> Result<Packet<'frame>>
     {
         let icmp_packet = Icmpv4Packet::new_checked(ip_payload)?;
         let checksum_caps = self.device_capabilities.checksum.clone();
-        let icmp_repr = Icmpv4Repr::parse(&icmp_packet, &checksum_caps)?;
+        let icmp_repr = Icmpv4Repr::parse(&icmp_packet, &checksum_caps)?.into();
```

Comment:
The new `IcmpRepr` really shouldn't be used in the `process_icmpv4` or `process_icmpv6` functions. AFAIK it is really only needed for getting data from the socket to the interface, but I don't think it is needed for parsing data from the outside world.

Reply:
It's needed for `IcmpSocket::send`.

Reply:
yeah this is more or less a nit. It might be cleaner to add a line like `let socket_repr = IcmpRepr::from(icmp_repr)` then use `socket_repr` for the `IcmpSocket` functions and then you can still `match icmp_repr` and you don't have the extra `IcmpRepr::Ipv4(...)`, but this is just a nit.

Reply:
You also need to add similar code to `process_icmpv6` to ensure any ICMPv6 sockets actually receive data.